rp2/machine_pin: Replace macros with Pico SDK functions.

Gadgetoid opened the pull request:
Replace custom macros with Pico SDK functions, enabling support for RP2350B variant chips with > 32 GPIOs.

I had missed this when porting to RP2350/RP2350B, since it only affects the string representation of a pin and has no functional implications (at least where users aren't inspecting that text to detect pin state).

### Summary

Fixes an issue raised in  #17241 where `machine_pin_print` was not reporting the correct pin direction for GPIOs >= 32, since the output states are stored in `gpio_oe_hi`

### Testing

A basic test on an RP2350B board (Pimoroni Pico Plus 2) to make sure `machine_pin_print` produces the correct string:

```python
>>> machine.Pin(31, machine.Pin.OUT)
Pin(GPIO31, mode=OUT)
>>> machine.Pin(33, machine.Pin.OUT)
Pin(GPIO33, mode=OUT)
>>> machine.Pin(33, machine.Pin.OUT, machine.Pin.PULL_UP)
Pin(GPIO33, mode=OUT, pull=PULL_UP)
>>> machine.Pin(33, machine.Pin.OUT, machine.Pin.PULL_DOWN)
Pin(GPIO33, mode=OUT, pull=PULL_DOWN)
```

And for funcsel:

```python
>>> a = machine.Pin(0)
>>> b = machine.Pin(1)
>>> i2c = machine.I2C(0, sda=a, scl=b)
>>> print(a)
Pin(GPIO0, mode=ALT, pull=PULL_UP, alt=I2C)
>>> print(b)
Pin(GPIO1, mode=ALT, pull=PULL_UP, alt=I2C)
```

And for open-drain `machine_pin_toggle`, observe the output enable bit being toggled:

```python
>>> od = machine.Pin(3, machine.Pin.OPEN_DRAIN)
>>> print(od)
Pin(GPIO3, mode=OPEN_DRAIN)
>>> od.toggle()
>>> f"{machine.mem32[0xd0000030] & 0xffffffff:32b}"
'10000000000000000000000000000000'
>>> od.toggle()
>>> f"{machine.mem32[0xd0000030] & 0xffffffff:32b}"
'10000000000000000000000000001000'
```

### Trade-offs and Alternatives

This is a low-impact change affecting (fixing) only the output of `machine_pin_print`. It might actually result in a small size increase (🤞) but is a better choice than trying to reimplement SDK functions and dealing with future repercussions.

Edit: I lied. This affects `machine_pin_toggle` too!

Gadgetoid committed:
rp2/machine_pin: Replace macros with Pico SDK functions.

Replace custom macros with Pico SDK functions, enabling support for RP2350B
variant chips with > 32 GPIOs.

Fixes #17241.

Signed-off-by: Phil Howard <github@gadgetoid.com>

Gadgetoid commented:
I think I need to run some more tests on open-drain toggling with pins >= 32 since I'm not seeing the register bits twiddled that I would expect.

Gadgetoid committed:
rp2: Fix simulated open drain for GPIOs >= 32.

machine_pin.c: Ensure uint64_t in GPIO_IS_OPEN_DRAIN macro.
mphalport.h: Ensure we're shifting a uint64_t.

Signed-off-by: Phil Howard <github@gadgetoid.com>

Gadgetoid commented:
Fixing the uint64_t mask handling for RP2350B support adds 80 bytes to the binary size on regular RP2.

We can probably set `machine_pin_open_drain_mask` to uint32_t on non-B variant builds and replace `1ULL` with a define that defaults to `1U` on non-B builds without too much hassle.

Gadgetoid committed:
rp2/machine_pin: Only use uint64_t for RP2350B variants.

machine_pin.c: Check NUM_BANK0_GPIOS and only use a uint64_t for
	       machine_pin_open_drain_mask when we have > 32 GPIOs.
mphalport.h: Match the behaviour in machine_pin.c, only using uint64_t
             where strictly necessary.

Signed-off-by: Phil Howard <github@gadgetoid.com>

Gadgetoid commented:
Commit 1: +16 bytes
Commit 2: +96 bytes
Commit 3: -132 bytes

Did a bit better than free!

dpgeorge commented on ports/rp2/machine_pin.c:
```diff
@@ -89,7 +83,13 @@ static const mp_irq_methods_t machine_pin_irq_methods;
 // Mask with "1" indicating that the corresponding pin is in simulated open-drain mode.
+#if NUM_BANK0_GPIOS > 32
 uint64_t machine_pin_open_drain_mask;
+#define MACHINE_PIN_OD_BIT 1ULL
```
I don't think these defines need to be redefined here, they should be picked up from `mphalport.h`.

Gadgetoid replied:
Fixed, thanks.

I do wonder if this would perhaps be better as a much more explicit set of macros:

```c
// Open drain behaviour is simulated.
#if NUM_BANK0_GPIOS > 32
extern uint64_t machine_pin_open_drain_mask;
#define MACHINE_PIN_OD_MASK(pin) (1ULL << pin)
#define GPIO_IS_OPEN_DRAIN(id) (machine_pin_open_drain_mask & (1ULL << id))
#define GPIO_SET_OPEN_DRAIN(id) (machine_pin_open_drain_mask |= (1ULL << id))
#define GPIO_CLR_OPEN_DRAIN(id) (machine_pin_open_drain_mask &= ~(1ULL << id))
#else
extern uint32_t machine_pin_open_drain_mask;
#define MACHINE_PIN_OD_MASK(pin) (1U << pin)
#define GPIO_IS_OPEN_DRAIN(id) (machine_pin_open_drain_mask & (1U << id))
#define GPIO_SET_OPEN_DRAIN(id) (machine_pin_open_drain_mask |= (1U << id))
#define GPIO_CLR_OPEN_DRAIN(id) (machine_pin_open_drain_mask &= ~(1U << id))
#endif
```

For instance this turns:

```c
static inline void mp_hal_pin_input(mp_hal_pin_obj_t pin) {
    gpio_set_dir(pin, GPIO_IN);
    machine_pin_open_drain_mask &= ~(1ULL << pin);
    gpio_set_function(pin, GPIO_FUNC_SIO);
}
```

into:

```c
static inline void mp_hal_pin_input(mp_hal_pin_obj_t pin) {
    gpio_set_dir(pin, GPIO_IN);
    GPIO_CLR_OPEN_DRAIN(pin);
    gpio_set_function(pin, GPIO_FUNC_SIO);
}
```

Gadgetoid committed:
rp2/machine_pin: Refactor open drain mask macros.

machine_pin.c: Move GPIO_IS_OPEN_DRAIN to mphalport.h.
               Use new GPIO_DISABLE_OPEN_DRAIN macro.
mphalport.h: Add GPIO_ENABLE/DISABLE_OPEN_DRAIN macros.

Signed-off-by: Phil Howard <github@gadgetoid.com>

dpgeorge reviewed (approved):
Looks good now, a nice refactor!
